docs: add docs for synthetic keys (#5518)
confluentinc/ksql: docs/developer-guide/joins/synthetic-keys.md

keywords: ksqldb, join, key, rowkey
---

Some joins have a synthetic key column in their result. This is a column that does not come from any

> source. Lets start with an example to help explain what these synthetic key columns are and why they

Reviewer: ```suggestion
source. Here's an example to help explain what synthetic key columns are and why they
```

are required:

```sql
CREATE TABLE OUTPUT AS
  SELECT * FROM L FULL OUTER JOIN R ON L.ID = R.ID;
```

> The above statement seems straight forward enough: create a new table that's the result of

Reviewer 2: straightforward

> The above statement seems straight forward enough: create a new table that's the result of

Reviewer: ```suggestion
The previous statement seems straightforward enough: create a new table that's the result of
```

> performing a full outer join of two source tables, joining on their ID columns. One might expect

Reviewer: ```suggestion
performing a full outer join of two source tables, joining on their ID columns. You might expect
```

that such a join would result in a table with a compound primary key containing both `L.ID` and
`R.ID`, and that's what will happen once ksqlDB supports
[structured keys][1].

> Unfortunately, until ksqlDB supports structured keys the join must result in a table with a single

Reviewer: ```suggestion
Unfortunately, until ksqlDB supports structured keys, the join must result in a table with a single
```

> primary key column. In the above join, as it is a full-outer join, either `L.ID` or `R.ID` may be

Reviewer: ```suggestion
primary key column. Because the previous join is a full-outer join, either `L.ID` or `R.ID` may be
```

> missing, i.e. `NULL`, or both may have the same value. However, the data produced to {{ site.aktm }}

Reviewer: ```suggestion
missing (`NULL`), or both may have the same value. But the data produced to {{ site.aktm }}
```

> will always have the message key set to the non-null `ID` column, as shown in the following table:

Reviewer: ```suggestion
always has the message key set to the non-null `ID` column, as shown in the following table:
```

| L.ID  | R.ID | Kafka message key |
|-------|------|:------------------|
|  10   | null | 10                |
|  null | 7    | 7                 |
|  8    | 8    | 8                 |

Clearly, the data stored in the {{ site.ak }} message's key does not match either of the source `ID`

> columns. Hence, it is a new column: a synthetic column, i.e. a column that does not belong to each 

Reviewer: ```suggestion
columns. Instead, it's a new column: a *synthetic* column, which means a column that doesn't belong to either 
```

source table.

## What joins result in synthetic key columns?

Any join where the key column in the result does not match any source column is said to have a 
synthetic key column.

The following types of joins result in a synthetic key column being added to the result schema:

1. `FULL OUTER` joins, for example:

    ```sql
   CREATE TABLE OUTPUT AS
      SELECT * FROM L FULL OUTER JOIN R ON L.ID = R.ID;
    ```

 
2. Any join where all expressions used in the join `ON` criteria are not simple column references.
   For example: 

    ```sql
   -- join on expressions other than column references:
   CREATE TABLE OUTPUT AS
      SELECT * FROM L JOIN R ON ABS(L.ID) = ABS(R.ID);
    ```

## What name is assigned to a Synthetic key column?

> The default name of a synthetic key column is `ROWKEY`. However, if any sources used in the join 

Reviewer: ```suggestion
The default name of a synthetic key column is `ROWKEY`. But if any sources used in the join 
```

> already contain a column called `ROWKEY` then the synthetic key column will be named `ROWKEY_1`, or

Reviewer: ```suggestion
already contain a column named `ROWKEY`, the synthetic key column is named `ROWKEY_1`, or
```
